rust/src/traits/samplers/bernoulli/mod.tex

```diff
@@ -0,0 +1,327 @@
+\end{theorem}
+\end{tcolorbox}
+
+\begin{proof}
+The function \texttt{sample\_geometric\_buffer} uses \texttt{fill\_bytes}  as a subroutine to generate a buffer of $B$ bytes where for each bit $b$ in the buffer it follows that $\Pr[b = 1] = \frac{1}{2}$ and $\Pr[b = 0] = \frac{1}{2}$ by Lemma 2.2. Conditioned on the event $E$, that there exists some bit in the buffer equal to $1$, the position of the \emph{first} such bit is a zero-indexed draw from the Geometric distribution $Geom(p)$ with $p = 0.5$ by definition of a Geometric random variable. If the event $E$ does not occur, the function returns \texttt{None}. 
+\end{proof}
+
+\subsection{Pseudocode \& proof of \texttt{sample\_bernoulli}}
+
+\begin{lstlisting}[language=Python, escapechar=|]
+# returns a single bit with some probability of success
+def sample_bernoulli(prob : T, constant_time : bool) -> bool:
+    if (prob < 0 or prob > 1): |\label{line:rangecheck}|
+        raise Exception("probability is not within [0, 1]")
+        
+            
+    if prob is 0: |\label{line:0check}|
+        return False
+        
+    if prob is 1: |\label{line:1check}|
+        return True
+        
+    # prepare for sampling first heads index by coin flipping
+    max_coin_flips = exact_int_cast(T::EXPONENT_BIAS) + exact_int_cast(T::MANTISSA_BITS) |\label{line:maxcoinflips}|
+    
+    # find number of bits to sample, rounding up to nearest byte (smallest sample size)
+    buffer_len = inf_div(max_coin_flips, 8) |\label{line:bufferlen}|
+    
+    # repeatedly flip fair coin and identify 0-based index of first heads
+    first_heads_index = sample_geometric_buffer(buffer_len, constant_time) |\label{line:sampling}|
+    
+    # if no events occurred, return early
+    if first_heads_index is None: |\label{line:noones}|
+        return False
+    
+    # find number of zeroes in binary rep. of prob
+    leading_zeroes = T::EXPONENT_BIAS - 1 - raw_exponent(prob) |\label{line:leadingzeroes}|
+    
+    match first_heads_index: |\label{line:match}|
+        # case 1: indexing into the leading zeroes
+        case first_heads_index < leading_zeroes: |\label{line:case1}|
+            return False
+        # case 2: indexing into implicit bit directly to left of mantissa
+        case first_heads_index == leading_zeroes: |\label{line:case2}|
+            return (raw_exponent(prob) != 0)
+        # case 3: indexing into out-of-bounds/implicitly-zero bits
+        case first_heads_index > leading_zeroes + MANTISSA_BITS: |\label{line:case3}|
+            return False
+        # case 4: indexing into mantissa
+        case _: |\label{line:case4}|
+            return (bin(prob) & (1 << MANTISSA_BITS + leading_zeroes - first_heads_index))
```

Comment:
The << operator has lower precedence than + and -, so the arithmetic expression needs parens. Whereas the parens next to the & are optional, since & has lower operator precedence than <<.